test/pprof.cc

```diff
@@ -0,0 +1,93 @@
+#include <catch2/catch.hpp>
+
+extern "C" {
+#include "native/pprof.h"
+}
+
+const char *empty_string = "";
+
+// cannot use null; use empty_string instead
+struct frame {
+  const char *function;
+  const char *filename;
+  int64_t lineno;
+  uint64_t function_id;
+  uint64_t location_id;
+};
+
+TEST_CASE("Basic PHP frame example", "[php]") {
+  // This case is specifically crafted to find a bug found when prototyping
+  // No bug found in shared library; bug was higher up
+  DProf dprof;
+
+  // set up the dprof obj
+  dprof.table_type = 1;
+  const char *sample_names[2] = {"samples", "wall-time"};
+  const char *sample_units[2] = {"count", "nanoseconds"};
+
+  static_assert(sizeof sample_names == sizeof sample_units,
+                "number of sample names and units must match");
+
+  std::size_t nsampletypes = sizeof sample_units / sizeof *sample_units;
+
+  REQUIRE(pprof_Init(&dprof, sample_names, sample_units, nsampletypes) == 0);
+
+  const char *executable_name = "php";
+  uint64_t mapping =
+      pprof_mapAdd(&dprof, 0, 0, 0, executable_name, empty_string);
+  REQUIRE(mapping == 1);
+
+  {
+    frame call_stack[4] = {
+        {"usleep", "", 0, 1, 1},
+        {"Time\\Sleep::usleep", "/home/circleci/app/profile.php", 23, 2, 2},
+        {"loop", "/home/circleci/app/profile.php", 37, 3, 3},
+        {"", "/home/circleci/app/profile.php", 45, 4, 4},
+    };
+
+    uint64_t locations[4];
+    uint64_t i = 0;
+    for (auto &frame : call_stack) {
+      auto function_id =
+          pprof_funAdd(&dprof, frame.function, empty_string, frame.filename, 0);
+      auto location_id =
+          pprof_locAdd(&dprof, mapping, 0, &function_id, &frame.lineno, 1);
+      REQUIRE(function_id == frame.function_id);
+      REQUIRE(location_id == frame.location_id);
+      locations[i++] = location_id;
+    }
+
+    int64_t values[2] = {1, 1};
+    REQUIRE(pprof_sampleAdd(&dprof, values, sizeof values / sizeof *values,
+                            locations, 4) == 0);
+  }
+
+  {
+    frame call_stack[4] = {
+        {"hrtime", "", 0, 5, 5},
+        {"Time\\SteadyClock::now", "/home/circleci/app/profile.php", 10, 6, 6},
+        {"loop", "/home/circleci/app/profile.php", 37, 3, 3},
+        {"", "/home/circleci/app/profile.php", 45, 4, 4},
+    };
+
+    uint64_t locations[4];
+    uint64_t i = 0;
+    for (auto &frame : call_stack) {
+      auto function_id =
+          pprof_funAdd(&dprof, frame.function, empty_string, frame.filename, 0);
+      auto location_id =
+          pprof_locAdd(&dprof, mapping, 0, &function_id, &frame.lineno, 1);
+      REQUIRE(function_id == frame.function_id);
+      REQUIRE(location_id == frame.location_id);
+      locations[i++] = location_id;
+    }
+
+    int64_t values[2] = {1, 1};
+    REQUIRE(pprof_sampleAdd(&dprof, values, sizeof values / sizeof *values,
+                            locations, 4) == 0);
+  }
+
+  // todo: what other properties should I assert?
+
+  REQUIRE(pprof_Free(&dprof) == 0);
```

Comment:
is it weird to you that the pprof_Free() function returns a value?
1.  Releasing a resource shouldn't error out.  I mean, obviously it can, and those situations are definitely bugs in the library--but the interface should ensure that the internal state is at least consistent enough to release.
2. If we can't make the assurance in 1 and it did throw an error, what could the developer possibly do about it?  I guess they could generate a new object and update all references, discarding the previous item as invalid.  Or--and this is basically the same thing--they could memset() over the struct and reuse it with new internal references (or the library could do this).  These DProf objects bind in only one direction, so failure would come from the release of a member rather than some kind of an invalid parent/global state (i.e., if an element were referenced through a global caching layer).

Reply:
Yeah, it's weird to me. I don't know what I would do if "free" failed. What do you do if a destructor in C++ fails? You abort, basically.

At the very least it needs signaling for what failure _means_. If it means an underlying call failed and I'm going to leak some memory, okay, thanks for letting me know. Maybe I can create a metric for it or disable the tracer on the next request or something. But without knowing what a failure means I couldn't really do anything.

Reply:
I think so too.  The implementation was a little broken, since it was safeguarding against a nonsensical error state (it was checking to see whether &dp->pprof is null, but that will never happen because struct DProf embeds a PPProfile).  This was a result of careless refactoring.

Fixed that issue and refactored to NULL for a later PR I'll send.